File: apps/ios/OuraApp/ModelCache.swift
#if TORCH
import CryptoKit
import Foundation

// Persisted per-unit results of the on-device models, so a reload after a sync
// only recomputes the days/nights whose inputs actually changed (usually just
// today) and a mid-run kill resumes instead of restarting. Entries are keyed by
// a fingerprint of the EXACT model inputs, so a cache hit is by construction the
// result the model would have produced — rebases, redecodes, and RingClock
// re-anchoring all change the inputs and therefore miss.

/// FNV-1a 64 — deterministic, dependency-free hashing of model input material.
struct FNV64 {
    private(set) var value: UInt64 = 0xcbf2_9ce4_8422_2325
    mutating func combine(_ x: UInt64) {
        var v = x
        for _ in 0..<8 {
            value = (value ^ (v & 0xff)) &* 0x0000_0100_0000_01b3
            v >>= 8
        }
    }
    mutating func combine(_ x: Int64) { combine(UInt64(bitPattern: x)) }
    mutating func combine(_ x: Int) { combine(Int64(x)) }
    mutating func combine(_ x: Double) { combine(x.bitPattern) }
    mutating func combine(_ x: Float) { combine(UInt64(x.bitPattern)) }
    mutating func combine(_ xs: [Float]) { combine(xs.count); for x in xs { combine(x) } }
    var hex: String { String(format: "%016llx", value) }
}

struct ActivityDayEntry: Codable {
    var fp: String
    var sessions: [WorkoutSession]
}

struct StagedNightEntry: Codable {
    var fp: String
    var stages: [Int]
}

private struct ModelCacheFile<Entry: Codable>: Codable {
    var version: Int
    var globalKey: String
    var entries: [String: Entry]
}

/// Mirrors SummaryCache: Application Support, serial queue, atomic writes.
enum ModelCacheStore {
    static let version = 2
    static let activityFile = "activity-model-cache.json"
    static let stagingFile = "sleep-staging-cache.json"
    private static let queue = DispatchQueue(label: "md.thomas.openoura.model-cache", qos: .utility)

    private static func url(_ file: String) -> URL {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent(file)
    }

    /// Entries for `file`, or empty when missing / from another schema version /
    /// written under a different global key (profile, timezone, DB, app build).
    static func load<E: Codable>(_ file: String, globalKey: String) -> [String: E] {
        guard let data = try? Data(contentsOf: url(file)),
              let decoded = try? JSONDecoder().decode(ModelCacheFile<E>.self, from: data),
              decoded.version == version, decoded.globalKey == globalKey
        else { return [:] }
        return decoded.entries
    }

    static func save<E: Codable>(_ file: String, globalKey: String, entries: [String: E]) {
        let payload = ModelCacheFile(version: version, globalKey: globalKey, entries: entries)
        queue.async {
            guard let data = try? JSONEncoder().encode(payload) else { return }
            try? data.write(to: url(file), options: .atomic)
        }
    }

    static func clearAll() {
        queue.sync {
            try? FileManager.default.removeItem(at: url(activityFile))
            try? FileManager.default.removeItem(at: url(stagingFile))
        }
    }

    /// Everything that invalidates every cached unit at once: model demographics,
    /// day bucketing, which DB is being read, and the app build (a shipped model
    /// or port change must not serve results from the old code).
    static func globalKey(profile: Profile?) -> String {
        let material = "v\(version)|\(profile?.sex ?? "")|\(profile?.age ?? -1)"
            + "|\(profile?.height_m ?? -1)|\(profile?.weight_kg ?? -1)|\(profile?.ring_size ?? -1)"
            + "|\(TimeZone.current.identifier)"
            + "|\(DB.readPath())"
            + "|\(Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "")"
        return SHA256.hash(data: Data(material.utf8)).map { String(format: "%02x", $0) }.joined()
    }
}
#endif

File: apps/android/app/src/full/java/md/thomas/openoura/models/ModelCache.kt
package md.thomas.openoura.models

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.KSerializer
import md.thomas.openoura.data.Profile
import md.thomas.openoura.data.SummaryJson
import md.thomas.openoura.data.WorkoutSession
import md.thomas.openoura.store.DB
import java.io.File
import java.security.MessageDigest
import java.util.TimeZone

// Persisted per-unit results of the on-device models, so a reload after a sync only
// recomputes the days/nights whose inputs actually changed (usually just today) and a
// mid-run kill resumes instead of restarting. Entries are keyed by a fingerprint of the
// EXACT model inputs, so a cache hit is by construction the result the model would have
// produced — rebases, redecodes, and RingClock re-anchoring all change the inputs and
// therefore miss. Port of apps/ios/OuraApp/ModelCache.swift.

/** FNV-1a 64 — deterministic, dependency-free hashing of model input material. */
class FNV64 {
    var value: ULong = 0xcbf2_9ce4_8422_2325uL
        private set

    fun combine(x: ULong) {
        var v = x
        repeat(8) {
            value = (value xor (v and 0xffuL)) * 0x0000_0100_0000_01b3uL
            v = v shr 8
        }
    }

    fun combine(x: Long) = combine(x.toULong())
    fun combine(x: Int) = combine(x.toLong())
    fun combine(x: Double) = combine(x.toRawBits().toULong())
    fun combine(x: Float) = combine(x.toRawBits().toLong().toULong())
    fun combine(xs: FloatArray) {
        combine(xs.size)
        for (x in xs) combine(x)
    }

    val hex: String get() = "%016x".format(value.toLong())
}

@Serializable
data class ActivityDayEntry(val fp: String, val sessions: List<WorkoutSession>)

@Serializable
data class StagedNightEntry(val fp: String, val stages: List<Int>)

@Serializable
private data class ModelCacheFile<E>(
    val version: Int,
    val globalKey: String,
    val entries: Map<String, E>,
)

/** Mirrors SummaryCache: app files dir, atomic writes. */
object ModelCacheStore {
    const val VERSION = 2
    const val ACTIVITY_FILE = "activity-model-cache.json"
    const val STAGING_FILE = "sleep-staging-cache.json"

    private fun file(context: Context, name: String) = File(context.filesDir, name)

    /**
     * Entries for [name], or empty when missing / from another schema version / written
     * under a different global key (profile, timezone, DB, app build).
     */
    fun <E> load(
        context: Context,
        name: String,
        globalKey: String,
        serializer: KSerializer<E>,
    ): MutableMap<String, E> = try {
        val f = file(context, name)
        if (!f.exists()) mutableMapOf() else {
            val decoded = SummaryJson.decodeFromString(
                ModelCacheFile.serializer(serializer), f.readText()
            )
            if (decoded.version != VERSION || decoded.globalKey != globalKey) mutableMapOf()
            else decoded.entries.toMutableMap()
        }
    } catch (_: Exception) {
        mutableMapOf()
    }

    fun <E> save(
        context: Context,
        name: String,
        globalKey: String,
        entries: Map<String, E>,
        serializer: KSerializer<E>,
    ) {
        try {
            val payload = ModelCacheFile(VERSION, globalKey, entries)
            val tmp = File(context.filesDir, "$name.tmp")
            tmp.writeText(
                SummaryJson.encodeToString(ModelCacheFile.serializer(serializer), payload)
            )
            tmp.renameTo(file(context, name))
        } catch (_: Exception) {
            // a cache write failure must never take down a model run
        }
    }

    fun clearAll(context: Context) {
        file(context, ACTIVITY_FILE).delete()
        file(context, STAGING_FILE).delete()
    }

    /**
     * Everything that invalidates every cached unit at once: model demographics, day
     * bucketing, which DB is being read, and the app build (a shipped model or port
     * change must not serve results from the old code).
     */
    fun globalKey(context: Context, profile: Profile?): String {
        val versionCode = try {
            context.packageManager.getPackageInfo(context.packageName, 0).longVersionCode
        } catch (_: Exception) {
            0L
        }
        val material = "v$VERSION|${profile?.sex ?: ""}|${profile?.age ?: -1}" +
            "|${profile?.heightM ?: -1}|${profile?.weightKg ?: -1}|${profile?.ringSize ?: -1}" +
            "|${TimeZone.getDefault().id}" +
            "|${DB.readPath(context)}" +
            "|$versionCode"
        return MessageDigest.getInstance("SHA-256")
            .digest(material.toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
    }
}
